platforms/native/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java

```diff
@@ -72,22 +72,23 @@ public abstract class AbstractNativeCompileTask extends DefaultTask {
     private final ConfigurableFileCollection includes;
     private final ConfigurableFileCollection systemIncludes;
     private final ConfigurableFileCollection source;
-    private final Map<String, String> macros = new LinkedHashMap<String, String>();
+    private final Map<String, String> macros = new LinkedHashMap<>();
     private final ListProperty<String> compilerArgs;
     private final IncrementalCompilerBuilder.IncrementalCompiler incrementalCompiler;
 
     public AbstractNativeCompileTask() {
-        ObjectFactory objectFactory = getProject().getObjects();
-        this.includes = getProject().files();
-        this.systemIncludes = getProject().files();
+        this.includes = getObjects().fileCollection();
+        this.systemIncludes = getObjects().fileCollection();
         dependsOn(includes);
         dependsOn(systemIncludes);
 
-        this.source = getTaskFileVarFactory().newInputFileCollection(this);
-        this.objectFileDir = objectFactory.directoryProperty();
-        this.compilerArgs = getProject().getObjects().listProperty(String.class);
-        this.targetPlatform = objectFactory.property(NativePlatform.class);
-        this.toolChain = objectFactory.property(NativeToolChain.class);
+        this.source = getObjects().fileCollection();
+        this.source.finalizeValueOnRead();
```

Comment:
❌ @big-guy Unless the bug with `finalizeValueOnRead()` where it discards task dependencies when read was fixed, this can cause issues. I don't think I had a chance to open the issue (or it was in our shared document), but when the file collection is realized, the task dependencies are dropped. I don't remember if it was a bug or intended behaviour. It would be worth checking that is someone early reads the source, everything works fine.

Reply:
We have some code on v7.6.4 that will exhibit the explained behaviour above. If it was fixed on 8.x, backporting this fix to 7.6.x will need to be adjusted.